File: WorkoutPlan.cpp
#include "WorkoutPlan.hpp"
#include <sstream>

PlanBase::PlanBase(const std::string& name) : planName(name) {}

PlanBase::PlanBase(const PlanBase& other) : planName(other.planName) {}

PlanBase::~PlanBase() = default;

std::string PlanBase::getPlanName() const {
    return planName;
}

WorkoutPlan::WorkoutPlan() : PlanBase("Workout Plan"), userProfile(nullptr) {}

WorkoutPlan::WorkoutPlan(const std::string& name) : PlanBase(name), userProfile(nullptr) {}

WorkoutPlan::WorkoutPlan(const WorkoutPlan& other)
    : PlanBase(other), upcomingWorkouts(other.upcomingWorkouts), 
      userProfile(other.userProfile ? std::make_unique<UserProfile>(*other.userProfile) : nullptr) {}

WorkoutPlan& WorkoutPlan::operator=(const WorkoutPlan& other) {
    if (this != &other) {
        PlanBase::operator=(other);
        upcomingWorkouts = other.upcomingWorkouts;
        if (other.userProfile) {
            userProfile = std::make_unique<UserProfile>(*other.userProfile);
        } else {
            userProfile.reset();
        }
    }
    return *this;
}

WorkoutPlan::WorkoutPlan(WorkoutPlan&& other) noexcept
    : PlanBase(std::move(other)), upcomingWorkouts(std::move(other.upcomingWorkouts)), 
      userProfile(std::move(other.userProfile)) {}

WorkoutPlan& WorkoutPlan::operator=(WorkoutPlan&& other) noexcept {
    if (this != &other) {
        PlanBase::operator=(std::move(other));
        upcomingWorkouts = std::move(other.upcomingWorkouts);
        userProfile = std::move(other.userProfile);
    }
    return *this;
}

WorkoutPlan::~WorkoutPlan() = default;

void WorkoutPlan::generatePersonalizedParams(const ExerciseCategory& category) {
    std::string params = "Params for " + category.getCategoryName();
    size_t pos = params.find("for");
    if (pos != std::string::npos) {
        params.replace(pos, 3, "based on");
    }
    (void)params;
}

void WorkoutPlan::adaptOnFeedback(const std::string& feedback) {
    std::stringstream ss;
    ss << "Adapted based on: " << feedback;
    std::string message = ss.str();
    message += "!";
}

std::vector<Workout> WorkoutPlan::getUpcoming(int count) const {
    std::vector<Workout> result;
    for (int i = 0; i < count && i < static_cast<int>(upcomingWorkouts.size()); ++i) {
        result.push_back(*upcomingWorkouts[i]);
    }
    return result;
}

bool WorkoutPlan::operator==(const WorkoutPlan& other) const {
    return getPlanName() == other.getPlanName();
}

bool WorkoutPlan::operator!=(const WorkoutPlan& other) const {
    return !(*this == other);
}

WorkoutPlan WorkoutPlan::operator+(const WorkoutPlan& other) const {
    WorkoutPlan combined(getPlanName() + " & " + other.getPlanName());
    combined.upcomingWorkouts.reserve(upcomingWorkouts.size() + other.upcomingWorkouts.size());
    combined.upcomingWorkouts.insert(combined.upcomingWorkouts.end(), upcomingWorkouts.begin(), upcomingWorkouts.end());
    combined.upcomingWorkouts.insert(combined.upcomingWorkouts.end(), other.upcomingWorkouts.begin(), other.upcomingWorkouts.end());
    return combined;
}

std::ostream& operator<<(std::ostream& os, const WorkoutPlan& plan) {
    os << "Workout Plan: " << plan.getPlanName() << "\nUpcoming: " << plan.upcomingWorkouts.size();
    return os;
}

File: WorkoutPlan.java
import java.util.ArrayList;
import java.util.List;

public class WorkoutPlan extends Workout {  // Придумываем наследование: WorkoutPlan extends Workout для демонстрации вызова базового конструктора
    private String planName;
    private UserProfile userProfile;
    private List<Workout> upcomingWorkouts = new ArrayList<>();

    public WorkoutPlan() {
        super();  // Вызов конструктора базового класса
        this.planName = "Workout Plan";
        this.userProfile = null;
    }

    public WorkoutPlan(String name) {
        super("Plan Date");  // Вызов конструктора базового класса с параметром
        this.planName = name;
        this.userProfile = null;
    }

    public WorkoutPlan(WorkoutPlan other) {
        super(other);  // Вызов копи-конструктора базового
        this.planName = other.planName;
        this.upcomingWorkouts = new ArrayList<>(other.upcomingWorkouts);
        this.userProfile = (other.userProfile != null) ? new UserProfile(other.userProfile) : null;
    }

    public WorkoutPlan assign(WorkoutPlan other) {
        if (this != other) {
            super.assign(other);  // Присваивание базового
            this.planName = other.planName;
            this.upcomingWorkouts = new ArrayList<>(other.upcomingWorkouts);
            this.userProfile = (other.userProfile != null) ? new UserProfile(other.userProfile) : null;
        }
        return this;
    }

    public String getPlanName() {
        return planName;
    }

    public void generatePersonalizedParams(ExerciseCategory category) {
        String params = "Params for " + category.getCategoryName();
        int pos = params.indexOf("for");
        if (pos != -1) {
            params = params.substring(0, pos) + "based on" + params.substring(pos + 3);
        }
    }

    public void adaptOnFeedback(String feedback) {
        StringBuilder sb = new StringBuilder();
        sb.append("Adapted based on: ").append(feedback);
        String message = sb.toString();
        message += "!";
    }

    public List<Workout> getUpcoming(int count) {
        List<Workout> result = new ArrayList<>();
        for (int i = 0; i < count && i < upcomingWorkouts.size(); i++) {
            result.add(new Workout(upcomingWorkouts.get(i)));
        }
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        WorkoutPlan other = (WorkoutPlan) obj;
        return planName.equals(other.planName);
    }

    public boolean notEquals(WorkoutPlan other) {
        return !this.equals(other);
    }

    public WorkoutPlan add(WorkoutPlan other) {
        WorkoutPlan combined = new WorkoutPlan(planName + " & " + other.planName);
        combined.upcomingWorkouts.addAll(upcomingWorkouts);
        combined.upcomingWorkouts.addAll(other.upcomingWorkouts);
        return combined;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Workout Plan: ").append(planName).append("\nUpcoming: ").append(upcomingWorkouts.size());
        return sb.toString();
    }
}